### PR TITLE
feat: Clickup action additions

### PR DESCRIPTION
**Pull request by ShayPunter**
## What does this PR do?

Adds the following action pieces:
- Get Task Comments
- Get Space
- Get Spaces
- Get Lists
- Create Folderless List

## Type of change

- [x] Chore (refactoring code, technical debt, workflow improvements)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### Get Task Comments
1. Make a new flow
2. Add ClickUp as action
3. Select 'Get Task Comments'
4. Enter task ID (which is obtainable through URL on ClickUp)
5. Should return something similar to this code block:
`  "body": {
    "comments": [
      {
        "id": "90100006324065",
        "comment": [
          {
            "text": "another one!",
            "attributes": {}
          },
          {
            "text": "\n",
            "attributes": {
              "block-id": "block-2b35326e-ba9c-4686-bf35-fba3a63bc248"
            }
          }
        ],
        "comment_text": "another one!\n",
        "user": {
          "id": 4544967,
          "username": "Shay Punter",
          "email": "[redacted]",
          "color": "#7b68ee",
          "initials": "SP",
          "profilePicture": "https://attachments.clickup.com/profilePictures/4544967_UFJ.jpg"
        },
        "reactions": [],
        "date": "1676742312882"
      },
      {
        "id": "90100006324064",
        "comment": [
          {
            "text": "this is coment 1",
            "attributes": {}
          },
          {
            "text": "\n",
            "attributes": {
              "block-id": "block-b0cc6218-4a57-46d4-82ff-403c28c80e02"
            }
          }
        ],
        "comment_text": "this is coment 1\n",
        "user": {
          "id": 4544967,
          "username": "Shay Punter",
          "email": "[redacted]",
          "color": "#7b68ee",
          "initials": "SP",
          "profilePicture": "https://attachments.clickup.com/profilePictures/4544967_UFJ.jpg"
        },
        "reactions": [],
        "date": "1676742311076"
      }`

### Get Space
1. Make a new flow
2. Add ClickUp as action
3. Select 'Get Space'
4. Enter Space ID (which is obtainable through URL on ClickUp (make sure to click on only a space))
5. Should return something similar to this code block:
`"body": {
    "id": "90100131481",
    "name": "activepieces",
    "color": null,
    "private": false,
    "avatar": null,
    "admin_can_manage": false,
    "statuses": [
      {
        "id": "p90100131481_MrBxr3Fs",
        "status": "to do",
        "type": "open",
        "orderindex": 0,
        "color": "#d3d3d3"
      },
      {
        "id": "p90100131481_0OngRj1i",
        "status": "complete",
        "type": "closed",
        "orderindex": 1,
        "color": "#6bc950"
      }
    ],
    "multiple_assignees": true,
    "features": {
      "due_dates": {
        "enabled": true,
        "start_date": true,
        "remap_due_dates": false,
        "remap_closed_due_date": false
      },
      "sprints": {
        "enabled": false
      },
      "time_tracking": {
        "enabled": true,
        "harvest": false,
        "rollup": false
      },
      "points": {
        "enabled": false
      },
      "custom_items": {
        "enabled": false
      },
      "priorities": {
        "enabled": true,
        "priorities": [
          {
            "color": "#f50000",
            "id": "1",
            "orderindex": "1",
            "priority": "urgent"
          },
          {
            "color": "#ffcc00",
            "id": "2",
            "orderindex": "2",
            "priority": "high"
          },
          {
            "color": "#6fddff",
            "id": "3",
            "orderindex": "3",
            "priority": "normal"
          },
          {
            "color": "#d8d8d8",
            "id": "4",
            "orderindex": "4",
            "priority": "low"
          }
        ]
      },
      "tags": {
        "enabled": true
      },
      "check_unresolved": {
        "enabled": true,
        "subtasks": null,
        "checklists": null,
        "comments": null
      },
      "zoom": {
        "enabled": false
      },
      "milestones": {
        "enabled": false
      },
      "custom_fields": {
        "enabled": true
      },
      "dependency_warning": {
        "enabled": true
      },
      "multiple_assignees": {
        "enabled": true
      }
    },
    "archived": false
  }`

### Get Spaces
1. Make a new flow
2. Add ClickUp as action
3. Select 'Get Spaces'
4. Enter Team ID (Is the first set of numbers in the URL)
5. Should return something similar to this code block:
`"body": {
    "spaces": [
      {
        "id": "49533063",
        "name": "Companies",
        "color": "#00D717",
        "private": true,
        "avatar": null,
        "admin_can_manage": true,
        "statuses": [
          {
            "id": "p49533063_p32430599_g1Z3tKPQ",
            "status": "to do",
            "type": "open",
            "orderindex": 0,
            "color": "#d3d3d3"
          },
          {
            "id": "p49533063_p32430599_vzwoLgck",
            "status": "complete",
            "type": "closed",
            "orderindex": 1,
            "color": "#6bc950"
          }
        ],
        "multiple_assignees": true,
        "features": {
          "due_dates": {
            "enabled": true,
            "start_date": true,
            "remap_due_dates": false,
            "remap_closed_due_date": false
          },
          "sprints": {
            "enabled": false
          },
          "time_tracking": {
            "enabled": true,
            "harvest": false,
            "rollup": false
          },
          "points": {
            "enabled": false
          },
          "custom_items": {
            "enabled": false
          },
          "priorities": {
            "enabled": true,
            "priorities": [
              {
                "color": "#f50000",
                "id": "1",
                "orderindex": "1",
                "priority": "urgent"
              },
              {
                "color": "#ffcc00",
                "id": "2",
                "orderindex": "2",
                "priority": "high"
              },
              {
                "color": "#6fddff",
                "id": "3",
                "orderindex": "3",
                "priority": "normal"
              },
              {
                "color": "#d8d8d8",
                "id": "4",
                "orderindex": "4",
                "priority": "low"
              }
            ]
          },
          "tags": {
            "enabled": true
          },
          "time_estimates": {
            "enabled": true,
            "rollup": false,
            "per_assignee": false
          },
          "check_unresolved": {
            "enabled": true,
            "subtasks": null,
            "checklists": null,
            "comments": null
          },
          "zoom": {
            "enabled": true
          },
          "milestones": {
            "enabled": true
          },
          "custom_fields": {
            "enabled": true
          },
          "dependency_warning": {
            "enabled": true
          },
          "multiple_assignees": {
            "enabled": true
          },
          "emails": {
            "enabled": true
          }
        },
        "archived": false,
        "members": [
          {
            "user": {
              "id": 4544967,
              "username": "Shay Punter",
              "color": "#7b68ee",
              "profilePicture": "https://attachments.clickup.com/profilePictures/4544967_UFJ.jpg",
              "initials": "SP"
            }
          }
        ]
      },
      {
        "id": "38651852",
        "name": "Shay Punter", (truncated)`

### Get List
1. Make a new flow
2. Add ClickUp as action
3. Select 'Get List'
4. Enter Team ID (Click on a list, ID is the last set of numbers without query strings)
5. Should return something similar to this code block:
`"body": {
    "id": "901000754585",
    "name": "List",
    "deleted": false,
    "orderindex": 0,
    "content": "",
    "priority": null,
    "assignee": null,
    "due_date": null,
    "start_date": null,
    "folder": {
      "id": "90100421496",
      "name": "hidden",
      "hidden": true,
      "access": true
    },
    "space": {
      "id": "90100131481",
      "name": "activepieces",
      "access": true
    },
    "inbound_address": "a.t.901000754585.u-4544967.93e9c256-ce9f-4101-a5d0-cf2b09eeb2e7@tasks.clickup.com",
    "archived": false,
    "override_statuses": false,
    "statuses": [
      {
        "id": "p90100131481_MrBxr3Fs",
        "status": "to do",
        "orderindex": 0,
        "color": "#d3d3d3",
        "type": "open"
      },
      {
        "id": "p90100131481_0OngRj1i",
        "status": "complete",
        "orderindex": 1,
        "color": "#6bc950",
        "type": "closed"
      }
    ],
    "permission_level": "create"
  }`
  
### Create Folderless List
1. Make a new flow
2. Add ClickUp as action
3. Select 'Create Folderless List'
4. Select Workspace 
5. Select Space
6. Enter name for new list
8. Should return something similar to this code block & new list should be created in the relevant area on ClickUp:
`"body": {
    "id": "901000755248",
    "name": "Task from AP",
    "deleted": false,
    "orderindex": 0,
    "content": "",
    "priority": null,
    "assignee": null,
    "due_date": null,
    "start_date": null,
    "folder": {
      "id": "90100421924",
      "name": "hidden",
      "hidden": true,
      "access": true
    },
    "space": {
      "id": "90100131481",
      "name": "activepieces",
      "access": true
    },
    "inbound_address": "a.t.901000755248.u-4544967.5fe59acb-3e97-4222-bfc0-f92d34a1bae6@tasks.clickup.com",
    "archived": false,
    "override_statuses": false,
    "statuses": [
      {
        "id": "p90100131481_MrBxr3Fs",
        "status": "to do",
        "orderindex": 0,
        "color": "#d3d3d3",
        "type": "open"
      },
      {
        "id": "p90100131481_0OngRj1i",
        "status": "complete",
        "orderindex": 1,
        "color": "#6bc950",
        "type": "closed"
      }
    ],
    "permission_level": "create"
  }`

## Checklist

### CODE REVIEW

**Comment by abuaboud** [comment]
@allcontributors please add ShayPunter for plugin

**Comment by abuaboud** [comment]
The PR is great, but the actions that involve reading data (such as getting task comments, getting a list, getting a space, and getting spaces) may not be very useful, as they fetch data that I am uncertain about they can be used.

Assuming that you intend to use the data in subsequent steps, I recommend replacing the properties in the next step with a dropdown property, that fetch the options from their accounts (check create-task.ts) for the reference.

I took look at most popular usecases, If we could have these actions in this PR that would be amazing.

1- Create Subtask
2- Post Comment Task
3- Create Folderless List (Already done)
 
@ShayPunter  What do you think?

I also created this Issue https://github.com/activepieces/activepieces/issues/644 for this PR.

File: packages/pieces/apps/src/lib/clickup/actions/comments/get-task-comments.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+		authentication: clickupCommon.authentication,
+		task_id: Property.ShortText({
+			description: 'The ID of the task to get',
+			displayName: 'Task ID',
+			required: true,
+		}),
+	},
+	async run(configValue) {
+		const { task_id, authentication } = configValue.propsValue;
+		return await callClickUpApi(HttpMethod.GET,
```
Comment (abuaboud) [inline packages/pieces/apps/src/lib/clickup/actions/comments/get-task-comments.ts]: I think it's better return body only, this applies for all others actions as well.

If there is an error, headers, body, status will be returned by activepieces http wrapper.

File: packages/pieces/apps/src/lib/clickup/actions/spaces/get-spaces.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import { clickupCommon, callClickUpApi } from "../../common";
+
+
+export const getClickupSpaces = createAction({
+	name: 'get_spaces',
+	description: 'Gets spaces in a ClickUp workspace',
+	displayName: 'Get Spaces',
+	props: {
+		authentication: clickupCommon.authentication,
+		team_id: Property.ShortText({
```
Comment (abuaboud) [inline packages/pieces/apps/src/lib/clickup/actions/spaces/get-spaces.ts]: The team_id refers to workspace_id as well in clickup, we can use the dropdown `workspace_id` in common file here instead.

File: docs/pieces/apps/clickup.mdx [hunk 3]
```diff
@@ -16,5 +16,45 @@ title: 'Clickup'
...
+      <Card title="Get Spaces">
+        Get all spaces in a ClickUp workspace
+      </Card>
+      <Card title="Get Task Comments">
+        Get all comments on a ClickUp task
+      </Card>
     </CardGroup>
     
+
+## How to get a list ID
```
Comment (abuaboud) [inline docs/pieces/apps/clickup.mdx]: Sorry for not mentioning that in the docs, but these docs are auto generated from piece folder.

You can create clickup.mdx in same directory and add the custom part below.

Example: https://github.com/activepieces/activepieces/blob/main/packages/pieces/apps/src/lib/discord/discord.mdx

**Review by abuaboud — CHANGES_REQUESTED** [review]
I learnt a lot from this PR, I will take an action and improve documentation / planning from our side on stuff I learnt!

File: packages/pieces/apps/src/lib/clickup/actions/comments/get-task-comments.ts (hunk 1 above)
Comment (ShayPunter) [reply packages/pieces/apps/src/lib/clickup/actions/comments/get-task-comments.ts]: Yep I agree with this, will update accordingly :)

**Review by abuaboud — APPROVED** [review]

**Comment by abuaboud** [comment]
This PR is elegant!

**Pull request merged**